### DISCUSSION TITLE
Update sample `web-client`

### DISCUSSION THREAD

**Discussion by AiyionPrime** [locka99/opcua#367] [Q&A]
@einarmo I drafted an update of actix in the sample web-server in https://github.com/vorausrobotik/opcua/commit/bc2ec95287259665a229c3e08658446529ab9624.

Currently this serves the webpage properly, but crashes upon connection through a websocket, as the `rt.block_on` [calls](https://github.com/vorausrobotik/opcua/commit/bc2ec95287259665a229c3e08658446529ab9624#diff-db3b4792a5fd696be65099ed9b7bcc17f47d011bae2e1ac7995f0efe9c803242R174-R183) are forbidden in the thread which handles async tasks.

```
WS: Ok(Text("connect opc.tcp://127.0.0.1:4855"))
thread 'actix-rt|system:0|arbiter:1' panicked at /home/aiyion/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.1/src/runtime/scheduler/multi_thread/mod.rs:86:9:
Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.
```

spawning a thread appears to not be that easy either, as `ctx` is `&mut` which cannot be sent through threads.

Do you have an idea for a different approach?

The draft @antiblue referenced in #271 is older than your additions, I think.

**Comment by einarmo** [comment]
Hm. It's quite unpleasant isn't it, I don't see an easy way to fix this, though I don't see how this didn't happen before my changes as well, there has always been a blocking runtime.

It doesn't actually seem like any uses of `block_on` actually _need_ the return value. A solution would be to use an `unbounded_channel` from tokio, since you can send messages on that without blocking, that way you could offload all the actual asynchronous processing to an event loop on a separate tokio task.

**Reply by einarmo** [reply]
That will require even more changes though, I can see, you'd need some way to communicate back with the actor... I don't know enough about actix to know the best way to do this.

**Reply by AiyionPrime** [reply]
That's a bummer, but thanks for the fast replies :)

@locka99 Do you have any suggestions how we go on with this sample?

**Reply by AiyionPrime** [reply]
DataNotifications did still work in 0.11.
Starting a bisect.
![Screenshot 2024-07-24 at 17-12-25 Web Socket client](https://github.com/user-attachments/assets/7f77f3fc-aa88-48fe-b275-37c5d47ab4e0)

**Reply by AiyionPrime** [reply]
43afb6a9e7d47c2e315cd1af9ecd9b3126c95b38 was the first commit where this sample stopped working, as the response as `element.value.value` does not exist since then anymore, but its capitalized variant does. `element.value.Value` One drawback is that it does not hold the plain value, but `{"Type":6,"Body":755}`.

If that change would not have stopped the example, the async change of the client would have, as the subscription to e.g. `ns=2;s=v1` does not work for the same (unknown) reasons as the regular client 
like discussed in #324.

@locka99 Is dropping the sample temporarily an option?
Until someone fixes it?

In the current state its not helping that much, I'm afraid.

**Reply by AiyionPrime** [reply]
If it was, I drafted its temporary removal in #368.
